### PR TITLE
menuconfig: Make search more flexible and search prompts

### PR DESCRIPTION
**Pull request by ulfalizer**
This commit gets the following incremental search improvements in from
upstream:

  - 1d3db5de9b8c2 ("menuconfig: Add search with multiple search
    strings")

    This makes a search string like 'foo bar' match all symbol names
    that match both 'foo' and 'bar' (which can be regexes), regardless
    of the order in which they appear in the match. This is faster and
    more flexible than having to type a bunch of '.*'.

  - 9bf8fc6e6907e ("menuconfig: Add prompts to incremental search")

    This makes the incremental searcher search prompt strings as well as
    symbol names.

    The prompt is now displayed next to the symbol name in the list of
    matches as well.

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

Screenshot:

skarmbild fran 2018-05-11 00-46-31

### CODE REVIEW

**Comment by ulfalizer** [comment]
CC @carlescufi @SebastianBoe @lpereira

File: scripts/kconfig/menuconfig.py [hunk 1]
```diff
@@ -1433,6 +1437,39 @@ def select_prev_match():
             s, s_i, hscroll = _edit_text(c, s, s_i, hscroll,
                                          edit_box.getmaxyx()[1] - 2)
 
+_cached_search_strings = None
+
+def _search_strings():
```
Comment (lpereira) [inline scripts/kconfig/menuconfig.py]: You can rewrite this as:

```Python
def _search_strings(cached_search_strings=[]):
   if not cached_search_strings:
       # populate the cache
   return cached_search_strings
```

Comment (ulfalizer) [reply scripts/kconfig/menuconfig.py]: Clever... had never thought of that use for the "interview question" default argument behavior. :P

I prefer to get rid of scopes early, even if it's a bit of extra code, so I think I'll keep the rest as is. I did the other variant in `_style()`.

Comment (ulfalizer) [reply scripts/kconfig/menuconfig.py]: Heh... `_CACHED_NODE_STRINGS = ...`. Blatant bug there. Gets pretty short too, so I'll go for it.

Too much coding over the last few days. :P

**Comment by ulfalizer** [comment]
@lpereira 
Fixed. Thanks for teaching me a new trick. :ok_hand:

**Review by carlescufi — APPROVED** [review]
Tested locally

**Pull request merged**